packages/lit-dev-content/site/blog/2022-02-07-eleventy.md

```diff
@@ -0,0 +1,181 @@
+it takes to download, parse, and evaluate the JavaScript implementations of
+those components.
+
+For a deeper dive into the benefits of pre-rendering, check out *[Rendering on
+the Web
+(developers.google.com)](https://developers.google.com/web/updates/2019/02/rendering-on-the-web)*.
+
+## How does it work?
+
+### Lit SSR
+
+The plugin uses `@lit-labs/ssr` under the hood, which works by creating an
+instance of each component in a lightweight browser-like environment within the
+Eleventy Node process, and evaluating the element's template using a special
+version of Lit's `render` function. This special `render` function behaves much
+like the `render` function that Lit uses in the browser, except that it emits
+stream of strings, instead of writing to the DOM.
+
+For more information about `@lit-labs/ssr`, checkout its [GitHub
+repo](https://github.com/lit/lit/tree/main/packages/labs/ssr). We'll be
+expanding the documentation soon to help developers integrate Lit pre-rendering
+into their tools and frameworks too.
+
+### Declarative Shadow DOM
+
+When a component is statically pre-rendered, its shadow DOM and styles are
+represented using the `<template shadowroot>` HTML element. When the browser's
+HTML parser encounters one of these elements, it applies the template's content
+as the shadow root of its parent element. This way, all of the DOM and style
+encapsulation guarantees of web components are preserved, all using declarative
+HTML.
+
+As of February 2022, Chrome and Edge have native support for Declarative Shadow
+DOM, but Firefox and Safari have not yet implemented it. A very small polyfill
+is available for browsers that don't yet have support. Integrating the polyfill
+into your site is
+[documented](https://github.com/lit/lit/tree/main/packages/labs/eleventy-plugin-lit#ponyfill)
+in the `@lit-labs/eleventy-plugin-lit` README.
+
+For more information about Declarative Shadow DOM, check out *[Declarative
+Shadow DOM (web.dev)](https://web.dev/declarative-shadow-dom/)*.
+
+### Re-hydration
+
+While static pre-rendering lets your components initial paint happen as quickly
+as possible, the components won't yet be interactive, because their JavaScript
+implementations haven't loaded.
+
+*Re-hydration* is the process where a statically pre-rendered component is
+upgraded to its JavaScript implementation, becoming responsive and interactive.
+
+Lit comes with automatic support for re-hydration via the special
+`lit/experimental-hydrate-support.js` module. As long as this module has been
+loaded, Lit components will detect when they have been pre-rendered, and adopt
+the shadow root that has already been attached to them.
+
+Re-hydration with Lit is very efficient, because Lit knows that it doesn't need
+to re-render the existing shadow root when it upgrades. If any data has changed
+between pre-rendering and upgrade, then only the parts of the shadow root that
+changed will be updated.
```

Comment:
This isn't entirely true today because we require the initial hydration to be done with the same data rendered on the server, after which the data can change and re-render (for now in the 11ty plugin this will always be true since the data can only come from static attributes).

This is a design choice, between hydration just priming the internal storage used for dirty-checking subsequent updates with initial data, vs. actually diffing what's in the DOM against the initial data and making changes at hydration time (we only do the former today). The open question of what we actually want to support is documented in https://github.com/lit/lit/issues/1434.

Reply:
For the last sentence, I think you could just say:
```suggestion
to re-render the existing shadow root when it upgrades. If data changes after
hydration, then only the parts of the shadow root that changed will be updated.
```